### ISSUE TITLE
Build Git emitted "error C1083:  Cannot open include file: 'openssl/ssl.h'" on MSVC

### ISSUE THREAD

**Issue by QuellaZhang** [git-for-windows/git#2217]
Setup
•Which version of Git for Windows are you using? Is it 32-bit or 64-bit?
 64-bit
 $ git --version --build-options
 git version 2.9.2.windows.1
 sizeof-long: 4
•Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?
 Windows Server 2016 64-bit + VS2017 15.7.2
 $ cmd.exe /c ver
 Microsoft Windows [Version 10.0.17134]

Description:
Git failed due to error C1083 when built, after our investigation, we found that "openssl/ssl.h" was not generated when it was used. So we removed "/m" option try to use a single thread to build, the error C1083 is gone, but we encountered another error "error MSB3030: Could not copy the file "..\\git-remote-http.exe" because it was not found.", the "git-remote-http.exe" still is not generated when it was used, can you help look at this? I provided logs.

 **Repro steps:**
 1.open VS2017 x64 Native Tools command
 2.git.exe clone -b vs/master https://github.com/git-for-windows/git D:\git\src
 3.cd D:\git\src
 4.git checkout "latest version"
 5.devenv /upgrade git.sln
 6.nuget.exe restore git.sln(optional)
 7.msbuild **/m** /p:Configuration=Release;Platform=x64 /p:WindowsTargetPlatformVersion=10.0.17134.0 /t:Rebuild git.sln

Failures:
[singlethread_log_x64_build.log](https://github.com/git-for-windows/git/files/3260357/singlethread_log_x64_build.log)
[multithreading_log_x64_build.log](https://github.com/git-for-windows/git/files/3260358/multithreading_log_x64_build.log)

D:\Git\src\compat\mingw.h(178,10): error C1083:  Cannot open include file: 'openssl/ssl.h': No such file or directory (d:\agent\_work\3\s\src\vctools\Compiler\CxxFE\sl\p1\c\p0prepro.c:1722) [D:\Git\src\xdiff\lib\xdiff_lib.vcxproj]
D:\Git\src\compat\mingw.h(178,10): error C1083: #include <openssl/ssl.h> [D:\Git\src\xdiff\lib\xdiff_lib.vcxproj]
D:\Git\src\compat\mingw.h(178,10): error C1083:    

D:\Git\src\git\LinkOrCopyBuiltins.targets(124,5): error MSB3030: Could not copy the file "..\\git-remote-http.exe" because it was not found. [D:\Git\src\git\git.vcxproj]

**Comment by dscho** [comment]
>  Cannot open include file: 'openssl/ssl.h': No such file or directory

Can you find any `ssl.h` in the `compat\vcbuild\vcpkg` tree?

**Comment by QuellaZhang** [comment]
Yes, I can find the "ssl.h" when the build is complete. But when compiling "xemit.c", there is no any "ssl.h" file. "..\git-remote-http.exe" is the same situation.

**Comment by dscho** [comment]
Ah, then the dependencies might be set incorrectly. Does MSBuild respect the project hierarchy defined in the `.sln` file?

**Comment by QuellaZhang** [comment]
Yes, I think it does. We also use the same MSBuild to build other projects, no such situation. We first encountereed this issue is on 4c5e026, but I could not find a specific version.

**Comment by QuellaZhang** [comment]
Any update?

**Comment by dscho** [comment]
> Any update?

No. The MSVC builds in our Azure Pipeline continue to work, and I fail to find the time to investigate this further on my end.

**Comment by QuellaZhang** [comment]
Ok, there are 3 issue.

1. "error C1083:  Cannot open include file: 'openssl/ssl.h': No such file or directory": Removed "/m" option from build line. **Resolved.**
2. "fatal error LNK1181: cannot open input file 'libcurl.lib'", related another issue: https://github.com/git-for-windows/git/issues/2195. VCPKG team has changed libcurl_imp.lib back to libcurl.lib https://github.com/microsoft/vcpkg/commit/db0c6043dcee80bbdbea6a320525f5fcf0b7dbda. **Resolved**.
3.  "error MSB3030: Could not copy the file "..\git-remote-http.exe" because it was not found.",  I patch
([patch.txt](https://github.com/git-for-windows/git/files/3333421/patch.txt)) for this to disable copy command in "LinkOrCopyBuiltins.targets". **Just a temporary workaround for us.**

Seems that https://github.com/git-for-windows/git/commit/f8ffef94ab135b7424e86d779eab66581449018a#diff-102829179bdbb6c7259a8732d04b9ed6 generated LinkOrCopyBuiltins.targets. I don't know how you actually generate git.sln. The final fix still needs you to complete.

**Comment by dscho** [comment]
>     1. Removed "/m" option from build line. **Resolved.**

Is this really resolved, given that the project hierarchy is clearly indicated in the `.sln`?

>     2\. VCPKG team has changed libcurl_imp.lib back to libcurl.lib [microsoft/vcpkg@db0c604](https://github.com/microsoft/vcpkg/commit/db0c6043dcee80bbdbea6a320525f5fcf0b7dbda). **Resolved**.

How is this resolved without reverting 6190a04ee82afb2cd6840028199df515e4192947?

Also, I am really surprised that this did not break our MSVC builds in https://dev.azure.com/git-for-windows/git/_build?definitionId=17. We build using the pipeline artifacts from https://dev.azure.com/git/git/_build?definitionId=9, and the most recent build is only three days old. If you care to look into the `compat` artifact of that build, you will see that the file is still called `vcbuild/vcpkg/installed/x64-windows/lib/libcurl_imp.lib`.

>     3\. "error MSB3030: Could not copy the file "..\git-remote-http.exe" because it was not found.",  I patch
>         ([patch.txt](https://github.com/git-for-windows/git/files/3333421/patch.txt)) for this to disable copy command in "LinkOrCopyBuiltins.targets".

This is almost certainly incorrect. The `git-remote-http.exe` file needs to be built, or it is an incomplete build.

If you care to download the pipeline artifact called `msvc-artifacts` from https://dev.azure.com/git-for-windows/git/_build/results?buildId=38594&view=results, you will notice that the enclosed `artifacts.tar.gz` _does_ contain `git-remote-http.exe`. It is therefore built via `make MSVC=1`, and it should be built via the Visual Studio Solution, too:

- https://github.com/git-for-windows/git/blob/2eb8195f73eb7717675e930a06d35dff993c59ea/git.sln#L25
- https://github.com/git-for-windows/git/blob/vs/master/git-remote-http/git-remote-http.vcxproj

**Comment by rimrul** [comment]
>>2\. VCPKG team has changed libcurl_imp.lib back to libcurl.lib [microsoft/vcpkg@db0c604](https://github.com/microsoft/vcpkg/commit/db0c6043dcee80bbdbea6a320525f5fcf0b7dbda). **Resolved**.
>Also, I am really surprised that this did not break our MSVC builds in https://dev.azure.com/git-for-windows/git/_build?definitionId=17. We build using the pipeline artifacts from https://dev.azure.com/git/git/_build?definitionId=9, and the most recent build is only three days old.

Seeing as the linked commit is less than a day old, it's probably not included in the build artifacts from 3 days ago. It'll probably be included in the scheduled build on July 1st.

**Comment by QuellaZhang** [comment]
How is this resolved without reverting 6190a04?

From issue https://github.com/git-for-windows/git/issues/2195: "Thanks for your reply! We tried to apply a patch as the code you prodived in fixup! msvc: support building Git using MS Visual C++ . And re-build, it still failed to fatal error LNK1181: cannot open input file 'libcurl.lib'. Looks like it doesn't work. Thanks Lin" 

"fatal error LNK1181: cannot open input file 'libcurl.lib'" has never been resolved by 6190a04. This problem has always existed.
 
And I checked yesterday's build files, compat artifact uses **"D:\Git\src\compat\vcbuild\vcpkg\installed\x64-windows\lib\libcurl.lib"**. Hmm, seems that we use the latest VCPKG source code, but I just follow the above build steps. I haved updated the build step clearer. You can easily to reproduce this issue. It will need 20 mins. BTW, attached yesterday's build log [log_x64_build.log](https://github.com/git-for-windows/git/files/3338276/log_x64_build.log)

![Untitled](https://user-images.githubusercontent.com/36754348/60327106-9f2e6f80-99bd-11e9-866a-6d5b33213898.png)

For the "git-remote-http.exe", I didn't stop it from built, I just disabled it from copying. But you are right, it is an incomplete build, we lost "git-remote-https.exe" "git-remote-ftp.exe" "git-remote-ftps.exe". If you have better workaround and fixes, I am happly to have a try.

         Creating hard link to copy "..\\git-remote-http.exe" to "..\\git-remote-https.exe".
         Creating hard link to copy "..\\git-remote-http.exe" to "..\\git-remote-ftp.exe".
         Creating hard link to copy "..\\git-remote-http.exe" to "..\\git-remote-ftps.exe".

**Comment by dscho** [comment]
> How is this resolved without reverting [6190a04](https://github.com/git-for-windows/git/commit/6190a04ee82afb2cd6840028199df515e4192947)?

I guess I'll answer myself: it is resolved because I was smart enough in that patch to look for the `.lib` and use whatever is there: `libcurl.lib` or `libcurl_imp.lib`.

> You can easily to reproduce this issue.

No, I can't, because I mainly use Azure Pipelines to build via MSVC.

> For the "git-remote-http.exe", I didn't stop it from built, I just disabled it from copying.

Wait, are you saying that there _is_ a `git-remote-http.exe` in your build?

**Comment by QuellaZhang** [comment]
Yes, I double checked that we can find ”git-remote-http.exe“ when the build is complete.

**Comment by dscho** [comment]
> I double checked that we can find ”git-remote-http.exe“ when the build is complete.

Can you double-check whether you _really_ need to drop copying it to the three other names, then?

**Comment by QuellaZhang** [comment]
Aha, we use Git project as one of RWC testing to detect compiler regression, Instead of actually using it. We can keep this issue active and I will continue to use the patch until this issue is fixed.

**Comment by dscho** [comment]
>  I will continue to use the patch until this issue is fixed.

That's the point of my question: it does not seem as if anything needs to be fixed here. You reported that the file `git-remote-https.exe` is generated correctly.

I could imagine that MSBuild just does not respect the sub-project dependencies, in which case your change to prevent MSBuild from building in parallel should have fixed your problem _already_.

**Comment by dscho** [comment]
> Ah, then the dependencies might be set incorrectly.

That was indeed the case. A pity I did not get help to figure this out!
